contracts/MechMarketplace.sol

```diff
@@ -177,30 +175,40 @@ contract MechMarketplace is IErrorsMarketplace {
         );
     }
 
-    /// @dev Calculates payment for request delivery.
+    /// @dev Calculates payment and fee based on delivery rates and mech type.
     /// @param mech Delivery mech address.
-    /// @param payment Payment amount.
+    /// @param requestId Request Id.
+    /// @param deliveryRate Delivery rate.
+    /// @param mechPayment Mech payment.
+    /// @param marketplaceFee Marketplace fee.
     function _calculatePayment(
         address mech,
-        uint256 payment
+        uint256 requestId,
+        uint256 deliveryRate
     ) internal virtual returns (uint256 mechPayment, uint256 marketplaceFee) {
-        // This must never be zero
-        if (payment > 0) {
-            // Process payment
-            marketplaceFee = (payment * fee) / 10_000;
-            mechPayment = payment - marketplaceFee;
-
-            // Check for zero value
-            if (mechPayment == 0) {
-                revert ZeroValue();
-            }
+        // Get actual delivery rate
+        uint256 actualDeliveryRate = IMech(mech).getFinalizedDeliveryRate(requestId);
```

Comment:
I believe this is a required change - get the finalized delivery rate that is recorded for every mech.